File: java/MergeTwoSortedLists.java
/*
	problem 21
	Merge two sorted linked lists and return it as a new list. 
	The new list should be made by splicing together the nodes of the first two lists.
*/

public class MergeTwoSortedLists{
	static public void main(String[] args){

	}

	// Definition for singly-linked list.
	static public class ListNode {
		int val;
		ListNode next;
		ListNode(int x) { val = x; }
	}

	static public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
		ListNode beforeHead = new ListNode(0);
		ListNode cursor = beforeHead;

		while (l1 != null && l2 != null){
			if(l1.val < l2.val){
				cursor.next = l1;
				l1 = l1.next;
			}else{
				cursor.next = l2;
				l2 = l2.next;
			}
			cursor = cursor.next;
		}

		cursor.next = l1 != null ? l1 : l2;

		return beforeHead.next;
	}
}

File: pythom/MergeTwoSortedLists.py
# problem 21
# Merge two sorted linked lists and return it as a new list. 
# The new list should be made by splicing together the nodes of the first two lists.
# Example:
# Input: 1->2->4, 1->3->4
# Output: 1->1->2->3->4->4

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        beforeHead = ListNode(0)
        node = beforeHead

        while l1 is not None and l2 is not None:
            if l1.val < l2.val:
                node.next = l1
                l1 = l1.next
            else:
                node.next = l2
                l2 = l2.next
        
            node = node.next

        if l1 is not None:
            node.next = l1
        else:
            node.next = l2

        return beforeHead.next
